providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -507,13 +537,33 @@ static void *mlkem_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
             goto err;
     }
 
-    if (!ossl_mlkem768_generate_key(mkey->encoded_pubkey, NULL, &mkey->seckey,
-                                    mkey->mlkem_ctx)
-        || !ossl_mlkem768_public_from_private(&mkey->pubkey, &mkey->seckey))
-        goto err;
+    if (mkey->encoded_privkey == NULL) {
+        mkey->encoded_privkey = OPENSSL_malloc(ossl_mlkem768_PRIVATE_KEY_BYTES);
+        if (mkey->encoded_privkey == NULL) {
+            goto err;
+        }
+    }
 
-    mkey->seckey_initialized = 1;
-    mkey->pubkey_initialized = 1;
+    if (gctx->seed != NULL) {
+        debug_print("MLKEMKM generate keys from seed");
+        if (!ossl_mlkem768_private_key_from_seed(&mkey->privkey, gctx->seed,
+                                            MLKEM_SEED_BYTES, mkey->mlkem_ctx)
+            || !ossl_mlkem768_marshal_private_key(mkey->encoded_privkey,
+                                                  &mkey->privkey)
+            || !ossl_mlkem768_public_from_private(&mkey->pubkey, &mkey->privkey)
+            || !ossl_mlkem768_marshal_public_key(mkey->encoded_pubkey,
+                                                 &mkey->pubkey))
+            goto err;
+    } else {
+        debug_print("MLKEMKM generate random keys");
+        if (!ossl_mlkem768_generate_key(mkey->encoded_pubkey, gctx->seed,
+                                        &mkey->privkey, mkey->mlkem_ctx)
+            || ossl_mlkem768_marshal_private_key(mkey->encoded_privkey,
```

Comment:
"!" missing -- wrong failure condition and the reason for "evp_extra" test failiing.